tests/unit/onyxClearWebStorageTest.js

```diff
@@ -0,0 +1,62 @@
+import Storage from '../../lib/storage';
+
+const ONYX_KEYS = {
+    DEFAULT_KEY: 'defaultKey',
+};
+jest.mock('../../lib/storage');
+
+describe('Set data while storage is clearing', () => {
+    let connectionID;
+    let Onyx;
+
+    /** @type OnyxCache */
+    let cache;
+
+    beforeAll(() => {
+        Onyx = require('../../index').default;
+        jest.useRealTimers();
+
+        Onyx.init({
+            keys: ONYX_KEYS,
+            registerStorageEventListener: () => {},
+            initialKeyStates: {
+                [ONYX_KEYS.DEFAULT_KEY]: 'default',
+            },
+        });
+        const webStorageClear = Storage.clear;
+        Storage.clear = jest.fn(() => new Promise(resolve => setTimeout(resolve, 20))
+            .then(webStorageClear));
+    });
+
+    // Always use a "fresh" cache instance
+    beforeEach(() => {
+        cache = require('../../lib/OnyxCache').default;
+    });
+
+    afterEach(() => {
+        Onyx.disconnect(connectionID);
+        return Onyx.clear();
+    });
+
+    it('should store merged values when calling merge on a default key after clear', () => {
+        let defaultValue;
+        connectionID = Onyx.connect({
+            key: ONYX_KEYS.DEFAULT_KEY,
+            initWithStoredValues: false,
+            callback: val => defaultValue = val,
+        });
+        const afterClear = Onyx.clear();
+
+        // Merge just after clear, on the next tick
+        const afterMerge = new Promise(resolve => setTimeout(() => {
+            Onyx.merge(ONYX_KEYS.DEFAULT_KEY, 'merged')
+                .then(resolve);
+        }, 0));
```

Comment:
So we are trying to simulate what exactly...? I think it's important to walk through the order of operations to see if it matters. This intentions need to be explained better or simplified.

1. `Onyx.clear()` is called and calls triggers a `getAllKeys()`
2. -- nothing happens with `merge()` because we are arbitrarily waiting
3.  `getAllKeys()` returns and we set the cached values then schedules the keys changed call and `Storage.clear()`
4. `merge` can start now and calls `get()`
5. ... does it matter? and how is this different from just calling `Onyx.merge()` right after `Onyx.clear()`?